### PR TITLE
[378] JWT long shelf life (clientside)

### PR DESCRIPTION
**Pull request by defong**
### Context
JWT long shelf life

### Changes proposed in this pull request
JWT tokens will be invalid due to
- expiration time
- mismatch 
  - audience
  - issuer
  - subject
 
### Guidance to review
`EncodeService` belongs nested in `DecodeService` that is by designed as it has very little to no reason to live. 

This is the client side of the service, and consumes the JWT.

Should be reviewed alongside
see https://github.com/DFE-Digital/publish-teacher-training/pull/1429

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by defong** [comment]
[[1 day] JWT long shelf life](https://trello.com/c/N0SzJtAH/378-1-day-jwt-long-shelf-life)

File: app/services/jwt/decode_service.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
+module JWT
+  class DecodeService
```
Comment (darokel) [inline app/services/jwt/decode_service.rb:2]: The structure of this is a little confusing. Do these classes need to be split out?

Comment (defong) [reply app/services/jwt/decode_service.rb:2]: I did started it out as `EncodeService` & `DecodeService`...

> EncodeService belongs nested in DecodeService that is by designed as it has very little to no reason to live.

`EncodeService` was mainly introduced to support testing, and to support legacy `mcb` token generation, it more or less the mirror and the younger sibling of `DecodeService`.

ie in order to decode something needs to be encoded first,

Comment (darokel) [reply app/services/jwt/decode_service.rb:2]: Discussed offline.

**Review by darokel — APPROVED** [review]

**Review by dankmitchell — APPROVED** [review]

File: app/services/jwt/decode_service.rb (hunk 1 above)
Comment (gpeng) [reply app/services/jwt/decode_service.rb:2]: I don't understand what's going on here either.

File: app/services/jwt/decode_service.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            exp_leeway: 6.seconds.to_i,
+            **claims,
+        },
+      )
+
+      (payload, _algorithm) = decoded_token
+
+      payload.with_indifferent_access[:data]
+    end
+
```
Comment (gpeng) [inline app/services/jwt/decode_service.rb:93]: Extra whitespace 👾

File: lib/mcb.rb [hunk 3]
```diff
@@ -112,7 +112,7 @@ def self.generate_apiv2_token(email:, encoding:, secret: nil)
       raise "Secret not provided"
     end
 
-    JWT.encode(payload, secret, encoding)
+    JWT::DecodeService.encode(payload: payload, secret: secret, algorithm: encoding)
```
Comment (gpeng) [inline lib/mcb.rb:115]: This naming can't be correct. Why isn't there either an `EncodeService.call` and `DecodeService.call` or else something like `TokenService.encode` and `TokenService.decode`. The `call` pattern being what we're settling on.

File: spec/lib/mcb_spec.rb [hunk 4]
```diff
@@ -76,19 +76,20 @@
...
           email: email,
           encoding: encoding,
           secret: secret,
         )
-        expect(token).to eq JWT.encode({ email: email },
-                                       secret,
-                                       encoding)
+
+        expect(JWT::DecodeService.call(encoded_token: encoded_token,
+    secret: secret,
```
Comment (gpeng) [inline spec/lib/mcb_spec.rb:91]: Indenting is off here.

File: spec/services/jwt/decode_service_spec.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    before :each do
+      allow(Time).to receive(:zone)
+        .and_return(OpenStruct.new(now: now))
+    end
+
+    it "token values are equal" do
+      expect(subject).to match(payload)
+    end
+
+    describe "mismatch" do
```
Comment (gpeng) [inline spec/services/jwt/decode_service_spec.rb:40]: This should be a context

File: spec/services/jwt/decode_service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      shared_examples "mismatch" do |option, exception, value = option|
+        context "#{option} settings" do
+          let("encode_service_#{option}".to_sym) { value.to_s }
+          it "raised exception" do
```
Comment (gpeng) [inline spec/services/jwt/decode_service_spec.rb:44]: `raises an`

File: spec/services/jwt/decode_service_spec.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      end
+
+      include_examples "mismatch", :secret, JWT::VerificationError
+      include_examples "mismatch", :algorithm, JWT::IncorrectAlgorithm, "HS512"
+      include_examples "mismatch", :audience, JWT::InvalidAudError
+      include_examples "mismatch", :issuer, JWT::InvalidIssuerError
+      include_examples "mismatch", :subject, JWT::InvalidSubError
+    end
+
+    describe "expired token" do
```
Comment (gpeng) [inline spec/services/jwt/decode_service_spec.rb:57]: context

File: spec/services/jwt/decode_service_spec.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      let(:now) { Time.zone.now - (5.minutes + 1.second) }
+
+      it "raised exception" do
```
Comment (gpeng) [inline spec/services/jwt/decode_service_spec.rb:60]: `raises an`

**Review by gpeng — CHANGES_REQUESTED** [review]
Looks good but we need to sort the naming/organisation of the `DecodeService` out.

File: app/services/jwt/decode_service.rb (hunk 1 above)
Comment (defong) [reply app/services/jwt/decode_service.rb]: Moved to MCB side

File: lib/mcb.rb (hunk 3 above)
Comment (defong) [reply lib/mcb.rb]: Moved `EncodeService` to `MCB` side

File: spec/lib/mcb_spec.rb (hunk 4 above)
Comment (defong) [reply spec/lib/mcb_spec.rb]: Fixed

File: spec/services/jwt/decode_service_spec.rb (hunk 5 above)
Comment (defong) [reply spec/services/jwt/decode_service_spec.rb]: Fixed

File: spec/services/jwt/decode_service_spec.rb (hunk 6 above)
Comment (defong) [reply spec/services/jwt/decode_service_spec.rb]: Fixed

File: spec/services/jwt/decode_service_spec.rb (hunk 7 above)
Comment (defong) [reply spec/services/jwt/decode_service_spec.rb]: Fixed

File: spec/services/jwt/decode_service_spec.rb (hunk 8 above)
Comment (defong) [reply spec/services/jwt/decode_service_spec.rb]: Fixed

File: app/services/jwt/decode_service.rb (hunk 2 above)
Comment (defong) [reply app/services/jwt/decode_service.rb]: Fixed

**Review by gpeng — APPROVED** [review]

**Pull request merged**